lang_GENERIC/analyze/dataflow_tainting.ml

```diff
@@ -0,0 +1,159 @@
+(* Yoann Padioleau
+ *
+ * Copyright (C) 2020 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file license.txt.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * license.txt for more details.
+ *)
+open Common
+open Il
+
+module F = Il
+module D = Dataflow
+module VarMap = Dataflow.VarMap
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Tainting dataflow analysis.
+ *
+ * This is a very rudimentary tainting analysis. Just intraprocedural,
+ * very coarse grained (taint whole array/object).
+ * This is step1 for taint tracking support in semgrep.
+ *)
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type mapping = unit Dataflow.mapping
+
+module DataflowX = Dataflow.Make (struct
+  type node = F.node
+  type edge = F.edge
+  type flow = (node, edge) Ograph_extended.ograph_mutable
+  let short_string_of_node n = 
+        Meta_il.short_string_of_node_kind n.F.n
+end)
+
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+let str_of_name ((s, _tok), sid) =
+    spf "%s:%d" s sid
+
+let option_to_varmap = function
+  | None -> VarMap.empty
+  | Some lvar -> VarMap.singleton (str_of_name lvar) ()
+
+(*****************************************************************************)
+(* Transfer *)
+(*****************************************************************************)
+(* Not sure we can use the Gen/Kill framework here.
+ *)
+
+let union = 
+  Dataflow.varmap_union (fun () () -> ())
+let diff =
+  Dataflow.varmap_diff (fun () () -> ()) (fun () -> true)
+
+let (transfer:
```

Comment:
@aryx: this is the data-flow algorithm in https://en.wikipedia.org/wiki/Data-flow_analysis that you gave a brown bag talk on right? A couple of questions:
1) Under which condition is the `fixpoint` function terminating? In the pseudo-code, there is a while, but I am not able to see how this algorithm does terminate.
2) How does the CFG look like? Namely, output of Ilflow_build.cfg_of_stmts is passed to the taint analysis. is this an topologically ordered list of CFG nodes, where a node is a stmt block?

Reply:
1: In general a fixpoint terminates when there is "stability", meaning applying the transfer function does not improve anything (you've reached the fixpoint). If you look at dataflow.ml in the implementation of fixpoint_worker you'll see a recursive call, that's the equivalent of the while in the pseudo-code.
2: It's not an ordered list, it's just a graph. You can see it with pfff -cfg_generic or -cfg_il /path/to/file
(you need to have dot from graphviz and gv/ghostview on your machine though.